File: priv/javascript/lib/tuple.js
import Kernel from './kernel';

class Tuple {
  
  constructor(...args){
    this.values = Object.freeze(args);
  }

  get(index) {
    return this.values[index];
  }

  count() {
    return this.values.length;
  }

  [Symbol.iterator]() {
    return this.values[Symbol.iterator]();
  }

  toString() {
    var i, s = "";
    for (i = 0; i < this.values.length; i++) {
      if (s !== "") {
        s += ", ";
      }
      s += this.values[i].toString();
    }

    return "{" + s + "}";    
  }

  static to_string(tuple){
    return tuple.toString();
  };

  static delete_at(tuple, index){
    let new_list = [];

    for (var i = 0; i < tuple.count(); i++) {
      if(i !== index){
        new_list.push(tuple.get(i));
      }
    }

    return Kernel.SpecialForms.tuple.apply(null, new_list);
  };

  static duplicate(data, size){
    let array = [];

    for (var i = size - 1; i >= 0; i--) {
      array.push(data);
    }

    return Kernel.SpecialForms.tuple.apply(null, array);
  };

  static insert_at(tuple, index, term){
    let new_tuple = [];

    for (var i = 0; i <= tuple.count(); i++) {
      if(i === index){
        new_tuple.push(term);
        i++;
        new_tuple.push(tuple.get(i));
      }else{
        new_tuple.push(tuple.get(i));
      }
    }

    return Kernel.SpecialForms.tuple.apply(null, new_tuple);
  };

  static from_list(list){
    return Kernel.SpecialForms.tuple.apply(null, list);
  };

  static to_list(tuple){
    let new_list = [];

    for (var i = 0; i < tuple.count(); i++) {
      new_list.push(tuple.get(i));
    }

    return Kernel.SpecialForms.list(...new_list);
  };
}

export default Tuple;

File: priv/std_lib/tuple.ex
defmodule ElixirScript.Tuple do
  @moduledoc false  
  require JS

  def duplicate(data, size) do
    JS.new(Bootstrap.Core.Tuple, do_duplicate(data, size, []))
  end

  defp do_duplicate(_, 0, list) do
    list
  end

  defp do_duplicate(data, size, list) do
    do_duplicate(data, size - 1, list ++ [data])
  end

  def to_list(tuple) do
    tuple["value"]
  end

  def insert_at(tuple, index, value) do
    JS.new(Bootstrap.Core.Tuple, do_insert_at(tuple, index, value, 0, []))
  end

  defp do_insert_at(tuple, index, value, current_index, list) do
    if current_index == length(tuple) do
      list
    else
      list = case index == current_index do
        true ->
          list ++ [value, tuple.get(current_index)]
        false ->
          list ++ [tuple.get(current_index)]
      end

      do_insert_at(tuple, index, value, current_index + 1, list)
    end
  end

  def delete_at(tuple, index) do
    JS.new(Bootstrap.Core.Tuple, do_delete_at(tuple, index, 0, []))
  end

  defp do_delete_at(tuple, index, current_index, list) do
    if current_index == length(tuple) do
      list
    else
      list = case index == current_index do
        true ->
          list
        false ->
          list ++ [tuple.get(current_index)]
      end

      do_delete_at(tuple, index, current_index + 1, list)
    end
  end

  def append(tuple, value) do
    JS.new(Bootstrap.Core.Tuple, to_list(tuple) ++ [value])
  end

end
